Libproxy -- with or without options?

kraileth opened the pull request:
Here's the current state of my libproxy port. The application supports a whole lot of build options, most of which are not currently a thing for Ravenports due to missing dependencies. I've tried to add the options that make sense right now but after fighting with them for some time I'm running out of ideas on how to get this right. In fact I have reached the point where I consider to delete the options again and just submit a port for a "bare" version of the application.

The "minimal" variant (that disables all options) seems to work. Do you think that would be good enough or should I invest more time to fix the problems with the options currently available (the VALA option and the PY36 one actually work, but the rest is giving me trouble)? Unfortunately there's a lot of things that I don't understand:

1) I have defined three options in OPTIONS_STANDARD. According the build log all three are turned OFF. Chapter 9 in the wiki mentions that options not listed there are treated as OFF. This made me think that the ones that _are_ listed should be ON. What is the right method to activate them for the :standard package?

2) When I enable e.g. PERL524 in "minimal" for testing purposes, Perl 5.24 is installed and used. Turning on PERL526 instead draws in Perl 5.26. So far, so good. For some reason however, Python 3.6 is used if either PY27, PY35 or PY36 is enabled. I have no idea clue how that could happen. Do you have any idea?

3) I tried to set the CMAKE_ARG "-DPYTHON_SITEPKG_DIR=" to %%PREFIX%%/lib/python3.6/site-packages instead of /lib/python3.6/site-packages for example. However the variable is not getting substituted and a directory with the name of "%%PREFIX%%" is created in the staging dir... I currently circumvent this problem by checking for "lib" in the staging dir and moving it if it exists. This at least works but it's certainly not the preferable solution. How can I use a variable in this situation?

4) Is it possible to somehow use advanced conditions for option helpers? I'm turning Vala on or off depending on if the option for it is set. That works fine. But with Perl for example I could only prevent building the binding if _both_ Perl options are disabled. How would you do that with Ravenports? I currently disable them by default and enable Perl again if either Perl option is ON. Again this is not a clean solution at all. How would you solve this?

5) This is not the newest version of the software that is available. I started with the version that is in ports and then updated to the newest one - but that won't build. The fix may or may not be trivial, however I couldn't make it compile. In the end this port currently has a version that is at least one point release newer than what is in ports but not the most recent one.

If you can give me some hints I'm ready to work on it some more and hopefully iron out the problems. Otherwise I'd prefer to clean up the port to have no options and then submit that version so that I can move on to other ports. Which way would you prefer?

jrmarino commented:
for 1, see "OPT_ON, " https://github.com/jrmarino/Ravenports/wiki/Chapter-9

for 2: A port would not have separate PERL and PYTHON options.  You would have ONE option (E.g. PERL) and you would set "USES_ON=perl".  Each profile defines what the perl default is.  There's no reason to have an option to build to something else.  By the way, you can specify default explicitly with the ":perl_default" variant specifier but I don't see why you need that here.

for 3: these variables are already defined, see Mk/Uses/python.mk
If you have to set PYTHON2 / PYTHON3 you can set "files/special.mk" to write code that can set this based on values python.mk

for 4: I guess if you use perl, you get the binding.  Just add it to the manifest as conditional products.  what's the problem having the binding?

for 5: we need to start with the latest version available.

jrmarino commented:
for 3, special.mk is not needed.
```
-WITH_PYTHON{{PYTHON_MAJOR_VER}}:BOOL=ON 
```
should be enough

kraileth commented:
Ok, as soon as I find time for it, I'll try to improve the port.

About 5: I can totally understand that policy. But in this case it might mean that I can only commit this port for Linux (where I assume that the current version will just build) if I cannot figure out why it breaks on fbsd. I already have a working port for glib-networking in pipe, but I need to get this one done first.

jrmarino commented:
well, I think it's more likely that the port is on hold while trying to troubleshoot the building problem on freebsd.  Depends on the problem -- but requires an understanding of the problem and the options for a solution to make a real decision.

kraileth commented:
Alright. I'll try to address as many issues as possible. And even if a blocker should remain for the port, this should not be wasted time. Libproxy will be a dependency for a lot of other future ports and is effectively a must-have in the long run. We'll see.

kraileth commented:
This port is still not ready but definitely in better shape than before. I think that I have addressed the issues 1 to 4 but now I'm having a problem that testing :standard fails with this message:

### FATAL ERROR ###  libperl.so.5.26 is not in located in /lib, /usr/lib or within the RPATH/RUNPATH

The library is present in the build chroot but it lives in /raven/lib/perl5/5.26/mach/CORE/ - and I guess that this makes the build system unhappy. I haven't encountered this issue before. How could this be fixed?

jrmarino commented:
The perl module sets "-Wl,rpath,${PREFIX}/lib/perl5/${PERL_VER}/${PERL_ARCH}/CORE" in the ldflags, so what's apparently is going on is your makefile is overwriting or ignoring it.

In any case, you must link your executables with an rpath.  That's what this message is saying.
(you also get the message if the perl library isn't present, but it seems to be set as a buildrun dependency so it should be there.)

kraileth commented:
I see. Since CMake and I are not exactly friends, I'm going after the other remaining issue first and try to get the latest version to compile on fbsd (in the vague hope that ignoring the ldflags was a bug that might have been fixed upstream already).

jrmarino commented:
cmake is famous for stripping out rpath.
see 
-DCMAKE_SKIP_BUILD_RPATH:BOOL=YES (bucket_03/neovim)
or
-DCMAKE_INSTALL_RPATH:STRING="(use perl definitions from perl.mk here)" (bucket_4A/llvm50)

The neovim example should be enough though

kraileth commented:
Alright. I've updated the port to use the latest released version (which simply worked after not applying the patch from FreeBSD...). The Perl binding issue is now fixed as well; the CMake examples you proposed are not even needed - it wasn't the main library that was affected but only the one for Perl. I skimmed through the commit history of Libproxy on GitHub and found a commit (22c0fcf24126cfe058626063e97217b5ee2e48e2) that allows deciding if the binding should link explicitly against Libperl or not.

Both :standard and the :minimal variant build and pass the tests. From my point of view this port is now ok. You will probably see a few more issues or ask me to move some lines around, but I'm now pretty confident that this can be committed in the near future (yay!).

kraileth commented:
Using the variable for the Python sitelib dir is certainly preferable. (I totally need to get familiar with the various make fragment files sometime...)

jrmarino commented on bucket_56/libproxy/manifests/plist.single:
```diff
@@ -0,0 +1,11 @@
+lib/libproxy.so.1.0.0
+lib/pkgconfig/libproxy-1.0.pc
+share/cmake/Modules/Findlibproxy.cmake
+%%PERL-ON%%lib/perl5/site_perl/mach/%%PERL_VER%%/Net/Libproxy.pm
```
replace 
lib/perl5/site_perl/mach/%%PERL_VER%% 
with
%%SITE_ARCH%%
(I believe)

kraileth replied:
Yes this works (will be in the next commit).

jrmarino commented on bucket_56/libproxy/manifests/plist.single:
```diff
@@ -0,0 +1,11 @@
+share/cmake/Modules/Findlibproxy.cmake
+%%PERL-ON%%lib/perl5/site_perl/mach/%%PERL_VER%%/Net/Libproxy.pm
+%%PERL-ON%%lib/perl5/site_perl/mach/%%PERL_VER%%/auto/Net/Libproxy/Libproxy.so
+%%PYTHON-ON%%lib/python%%PYTHON_VER%%/site-packages/libproxy.py
```
replace
lib/python%%PYTHON_VER%%/site-packages
with
%%PYTHON_SITELIBDIR%%
(i believe)

jrmarino commented on bucket_56/libproxy/specification:
```diff
@@ -0,0 +1,66 @@
+
+DOWNLOAD_GROUPS=	main
+SITES[main]=		https://github.com/libproxy/libproxy/archive/
+DISTFILE[1]=		${PORTVERSION}.tar.gz:main
```
Intentional to use an archive tarball over a generated one (e.g. generated:main) ?

kraileth replied:
With the old version that I was original trying to port I somehow couldn't get this to work. I just tried again and now the generated distfile is all fine.

jrmarino commented on bucket_56/libproxy/specification:
```diff
@@ -0,0 +1,66 @@
+post-stage:
+	${STRIP_CMD} ${STAGEDIR}${PREFIX}/bin/proxy
+	${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/libproxy.so.*
+	test -f ${STAGEDIR}${PREFIX}/lib/perl5/site_perl/mach/${PERL_VER}/auto/Net/Libproxy/Libproxy.so && ${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/perl5/site_perl/mach/${PERL_VER}/auto/Net/Libproxy/Libproxy.so || true
```
1) why is libproxy.so not alway there?
2) you can replace this with 
```
-${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/perl5/site_perl/mach/${PERL_VER}/auto/Net/Libproxy/Libproxy.so
```
I think

jrmarino replied:
also, perl variable can be used to make this shorter

kraileth replied:
It will be missing if the Perl binding is not being built. I'll improve the command with your suggestions.

jrmarino commented on bucket_56/libproxy/specification:
```diff
@@ -0,0 +1,66 @@
+[VALA].CMAKE_ARGS_ON=			-DWITH_VALA:BOOL=ON
+[VALA].CMAKE_ARGS_OFF=			-DWITH_VALA:BOOL=OFF
+
+post-stage:
```
change
post-stage:
to
post-install:

kraileth commented:
The issues that you pointed at should have been addressed now.

jrmarino commented on bucket_56/libproxy/specification:
```diff
@@ -0,0 +1,66 @@
+post-stage:
+	${STRIP_CMD} ${STAGEDIR}${PREFIX}/bin/proxy
+	${STRIP_CMD} ${STAGEDIR}${PREFIX}/lib/libproxy.so.*
+	-${STRIP_CMD} ${STAGEDIR}${PREFIX}/${SITE_ARCH}/auto/Net/Libproxy/Libproxy.so
```
Now that I know why this is not always present, there's a better approach.
replace this with:
```

post-install-PERL-ON:
        ${STRIP_CMD} ${STAGEDIR}${PREFIX}/${SITE_ARCH}/auto/Net/Libproxy/Libproxy.so
```

kraileth commented:
Done.

jrmarino commented:
why is this port 3 versions behind?
0.4.12 when the latest release is 0.4.15
I found it because it fails to build on DF due to a typo that was fixed upstream months ago.

jrmarino commented:
nvm, i was looking at ports.  I guess the typo fix is post 0.4.15
